File: src/DotNetEveryDay.Extensions/Extensions/EnumerableExtensions.fs
namespace IdentityServer4.Helpers.Extensions

open System
open System.Collections.Generic
open System.Runtime.CompilerServices

[<Extension>]
type EnumerableExtensions =
    [<Extension>]
    static member inline IsNullOrEmpty(collection: seq<'T>) =
        collection = null || collection |> Seq.isEmpty        

    [<Extension>]
    static member inline HasDuplicates(collection: seq<'T>, selector: Func<'T, 'TProp>) : bool =
        if collection.IsNullOrEmpty() then
            false
        else
            let hashSet: HashSet<'TProp> = new HashSet<'TProp>()
            let convertedSelector = FuncConvert.FromFunc selector
        
            collection |> Seq.exists(convertedSelector >> hashSet.Add >> not)

File: src/DotNetEveryDay.Extensions/Extensions/EnumerableExtensions.cs
using System.Runtime.CompilerServices;

namespace DotNetEveryDay.Extensions.Extensions;

public static class EnumerableExtensions
{
    /// <summary>
    /// Check if collection is null or empty
    /// </summary>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool IsNullOrEmpty<T>(this IEnumerable<T>? enumerable)
        => enumerable is null || !enumerable.Any();

    /// <summary>
    /// Check <see cref="enumerable"/> duplicates by some compare field
    /// </summary>
    /// <param name="enumerable"></param>
    /// <param name="compareField"></param>
    /// <typeparam name="T"></typeparam>
    /// <typeparam name="TProp"></typeparam>
    /// <returns></returns>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool HasDuplicates<T, TProp>(this IEnumerable<T>? enumerable, Func<T, TProp> compareField)
    {
        if (enumerable is null)
            return false;

        var array = enumerable.ToArray();

        // TODO: [Optimization] Memory might be optimized
        var hashSet = new HashSet<TProp>(array.Length);
        
        foreach (var item in array)
            if (!hashSet.Add(compareField(item)))
                return true;

        return false;
    }
}